### ISSUE TITLE
Rule request: no-useless-builtin-calls

### ISSUE THREAD

**Issue by tbassetto** [eslint/eslint#7310]
**Please describe what the rule should do:**

The rule would prevent users from creating an `Object` of an object, a `Number` if a number, a `Boolean` of a boolean, etc.

Note that:
- [no-new-object](http://eslint.org/docs/rules/no-new-object) will not block code like `Object({a:1})` (without the `new`)
- [no-restricted-globals](http://eslint.org/docs/rules/no-restricted-globals) could work with a custom list of `Boolean`, `Object`, `Number` and `String` but it would block code like `a instanceof String`.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[X] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

``` js
const alreadyAnObject = new Object({a: 1});
const alreadyABoolean = new Boolean(true);
const alreadyAString = new String('hello world');
const alreadyANumber = new Number(12);
```

**Why should this rule be included in ESLint (instead of a plugin)?**

The 4 examples I gave are literally useless. What I propose is an improvement of [no-new-object](http://eslint.org/docs/rules/no-new-object), but more powerful and with a needed renaming.

**Comment by kaicataldo** [comment]
If we decide to implement this, I do think it makes sense to deprecate `no-new-object` and create a new rule that warns for all these patterns (with the ability to turn each type on/off).

**Comment by platinumazure** [comment]
Wait, does [no-new-wrappers](http://eslint.org/docs/rules/no-new-wrappers) cover this use case?

**Comment by kaicataldo** [comment]
@platinumazure Good call - I hadn't encountered that rule yet!

**Comment by ljharb** [comment]
@platinumazure it does not seem like it covers `Object({ anObjectLiteral })` (without the `new`)

**Comment by kaicataldo** [comment]
I'm curious why we have both `no-new-object` and `no-new-wrappers`. Seems like this could all be one rule.

**Comment by platinumazure** [comment]
@ljharb Correct.

@kaicataldo Maybe, but `no-new-object` also covers any use of `new Object()` when object literal syntax would do (in other words, I see it as more than enforcing object wrappers). I wouldn't have a problem folding `no-new-object`'s empty constructor check into `no-new-wrappers`, or some other sensible combination of the rules.

**Comment by tbassetto** [comment]
At the moment, both [`no-new-wrappers`](https://github.com/eslint/eslint/blob/master/lib/rules/no-new-wrappers.js) and  [`no-new-object`](https://github.com/eslint/eslint/blob/master/lib/rules/no-new-object.js) are very similar. They only check `node.callee.name` for `NewExpression` against a blacklist. That should be an argument to merge them, or to add a different behavior to one of the rule so that they are really different.

The code I have found in a project, which triggered the creation of this issue is the following:
```js
Object({a: 1});
String('232');
// etc.
```
Type "casting", but without the `new` keyword. And redundant (`Boolean('yes')` would actually do something interesting). None of the rules discussed about can prevent that.

**Comment by kaicataldo** [comment]
Thinking about this some more, I'm not sure combining `no-new-object` and `no-new-wrappers` is the best course here, since we also have `no-array-constructor`. They all do slightly different things, and I think it might be best to leave them as is (though we should maybe consider naming them in such a way that it's obvious they're related).

Putting the discussion to combine `no-new-object` and `no-new-wrappers` aside for a moment, disallowing unnecessary typecasting seems like a fundamentally different problem than what the `no-new-wrappers` rule is trying to solve (preventing unnecessary type coercion of a value vs. preventing creating wrappers objects rather than primitive values). 

That being said, that does seem like a useful rule - maybe something like `no-useless-coercion`? The rule would disallow type coercion with the following function calls if the argument is already of that type:
```js
Object({a: 1});
Boolean(true);
String('hello world');
Number(12);
```
Thoughts?

**Comment by tbassetto** [comment]
I obviously like the idea, and if the ESLint team think we should explore it I would like to create the PR for it :)

**Comment by kaicataldo** [comment]
@tbassetto Thanks for your willingness - makes this a ton easier if the team decides to go this route!

@eslint/eslint-team Meant to ping on my previous comment - thoughts?

**Comment by platinumazure** [comment]
@kaicataldo I like that idea and I've put a thumbs-up on your comment. We should probably clean up this issue (or create a new one) to reflect the concrete proposal, can you take the lead on that?

**Comment by kaicataldo** [comment]
@platinumazure Yeah, I can definitely make a new issue with a real proposal - just wanted to float the idea to see what others thought :)

**Comment by mikesherov** [comment]
An interesting aside is that `false` and `true` are not equivalent to `new Boolean(false)` and `new Boolean(true)`. Boolean objects allow properties, whereas literals do not. Famously used in Modernizr for their video support detection.

**Comment by ljharb** [comment]
That's true for all primitives but null and undefined (Symbol requires `Object()` ofc)

**Comment by kaicataldo** [comment]
Sorry, lost track of this. Will try to write up a real proposal this weekend.

**Comment by kaicataldo** [comment]
Created [a proposal](https://github.com/eslint/eslint/issues/7545) for the rule described in [this comment](https://github.com/eslint/eslint/issues/7310#issuecomment-253091809). If you feel this solves your issue, @tbassetto, feel free to close this issue. Are you still interested in helping to implement the rule if it's accepted?

**Comment by tbassetto** [comment]
If it's accepted, yes I am still interested.

**Comment by kaicataldo** [comment]
@tbassetto Awesome - thanks for your willingness
